frontend/src/routes/password.change.tsx

```diff
@@ -0,0 +1,293 @@
+
+const CURRENT_VIEWER_QUERY = graphql(/* GraphQL */ `
+  query CurrentViewerQuery {
+    viewer {
+      __typename
+      ... on Node {
+        id
+      }
+    }
+  }
+`);
+
+const CHANGE_PASSWORD_MUTATION = graphql(/* GraphQL */ `
+  mutation ChangePassword(
+    $userId: ID!
+    $oldPassword: String!
+    $newPassword: String!
+  ) {
+    setPassword(
+      input: {
+        userId: $userId
+        currentPassword: $oldPassword
+        newPassword: $newPassword
+      }
+    ) {
+      status
+    }
+  }
+`);
+
+export const Route = createFileRoute("/password/change")({
+  async loader({ context, abortController: { signal } }) {
+    const viewer = await context.client.query(
+      CURRENT_VIEWER_QUERY,
+      {},
+      { fetchOptions: { signal } },
+    );
+    if (viewer.error) throw viewer.error;
+    if (viewer.data?.viewer.__typename !== "User") throw notFound();
+  },
+
+  component: ChangePassword,
+});
+
+function ChangePassword(): React.ReactNode {
+  const { t } = useTranslation();
+  const [viewer] = useQuery({ query: CURRENT_VIEWER_QUERY });
+  if (viewer.error) throw viewer.error;
+  if (viewer.data?.viewer.__typename !== "User") throw notFound();
+  const userId = viewer.data.viewer.id;
+
+  const currentPasswordRef = useRef<HTMLInputElement>(null);
+  const newPasswordRef = useRef<HTMLInputElement>(null);
+  const newPasswordAgainRef = useRef<HTMLInputElement>(null);
+
+  const [result, changePassword] = useMutation(CHANGE_PASSWORD_MUTATION);
+
+  const onSubmit = (event: FormEvent<HTMLFormElement>): void => {
+    event.preventDefault();
+
+    const formData = new FormData(event.currentTarget);
+
+    const oldPassword = formData.get("current_password");
+    const newPassword = formData.get("new_password");
+    const newPasswordAgain = formData.get("new_password_again");
+
+    if (!oldPassword || !newPassword || !newPasswordAgain) {
+      throw new Error("not all passwords set");
+    }
+
+    // Assert that these are strings — according to the type system they could be Files
+    if (typeof oldPassword !== "string") {
+      throw new Error("oldPassword not string");
+    }
+    if (typeof newPassword !== "string") {
+      throw new Error("newPassword not string");
+    }
+
+    if (newPassword !== newPasswordAgain) {
+      throw new Error("passwords mismatch; this should be checked by the form");
+    }
+
+    changePassword({ userId, oldPassword, newPassword });
+  };
+
+  const success =
+    result.data && result.data.setPassword.status == SetPasswordStatus.Allowed;
+  const handleableError =
+    result.data && result.data.setPassword.status != SetPasswordStatus.Allowed;
+  const unhandleableError = result.error !== undefined;
+
+  const errorMsg: string | undefined = ((): string | undefined => {
+    switch (result.data?.setPassword.status) {
+      case SetPasswordStatus.InvalidNewPassword:
+        return t(
+          "frontend.password_change.failure.description_INVALID_NEW_PASSWORD",
```

Comment:
That would be the first time we'd use ALL CAPS in a translation key. I'd rather nest, like
```suggestion
          "frontend.password_change.failure.description.invalid_new_password",
```

Reply:
orginally I was building this dynamically from the content of the enum, but that upset quite a few assumptions that we only use static translation keys, so that's why this came about ;-)